main.tf

```diff
@@ -1,3 +1,28 @@
+locals {
+  origin_hostname_options = {
+    use_host = var.s3_origin_hostname != "" ? var.s3_origin_hostname : null
+    use_name = var.s3_origin_name != "" ? data.aws_s3_bucket.s3_origin[0].bucket_domain_name : null
+  }
+
+  origin_hostname        = local.origin_hostname_options[var.s3_origin_name != "" ? "use_name" : "use_host"]
+  override_origin_policy = var.override_s3_origin_policy && var.s3_origin_name != ""
+}
+
+# Workaround for the input variable validation
+resource "null_resource" "either_s3_origin_hostname_or_s3_origin_name_is_required" {
```

Comment:
Just out of curiosity, this approach did not work?
https://www.terraform.io/docs/language/values/variables.html#custom-validation-rules

Reply:
As far as I know the validation rules are only valid for the current variable scope. The other variable are not accessible.